### PR TITLE
[WIP] sbt 1.0

### PR DESCRIPTION
**Pull request by larsrh**
Ports tut to sbt 1.0

# Current state

* review by @SethTisue, @dwijnand & @stephennancekivell addressed
* submitted to `series/0.6.x` branch

# Open problems

* [ ] scripted test for 2.10 fails because an artifact is missing
* [ ] cross-build for 0.13.x and 1.0.x (I conjecture this isn't possible)

### CODE REVIEW

**Comment by tpolecat** [comment]
Awesome, I don't understand SBT enough to do any of these things. Thanks!

**Comment by larsrh** [comment]
Me neither, but you know, if it compiles …

**Comment by larsrh** [comment]
All tests pass, except for two:

* the "watch sources" test can't be implemented anymore due to sbt/io#67
* the "quick" test is failing, probably because I don't understand how `cached` works now

**Comment by larsrh** [comment]
This all works now _except_ for the Travis test for 2.10.x. It complains about a missing scripted for 2.10 but I'm not sure why it's trying to fetch it in the first place.

File: plugin/src/main/scala/tut/TutPlugin.scala [hunk 1]
```diff
@@ -40,15 +42,20 @@ object TutPlugin extends AutoPlugin {
...
+          },
+          NothingFilter
+        ),
+      tutFiles := tutFilesParser(false).value,
       scalacOptions in Tut := (scalacOptions in (Compile, console)).value,
       tutPluginJars := {
         // no idea if this is the right way to do this
         val deps = (libraryDependencies in Tut).value.filter(_.configurations.fold(false)(_.startsWith("plugin->")))
-        update.value.configuration("plugin").map(_.modules).getOrElse(Nil).filter { m =>
+        update.value.configuration(ConfigRef("plugin")).map(_.modules).getOrElse(Nil).filter { m =>
```
Comment (larsrh) [inline plugin/src/main/scala/tut/TutPlugin.scala]: This looks fishy.

File: plugin/src/main/scala/tut/TutPlugin.scala [hunk 2]
```diff
@@ -66,24 +73,20 @@ object TutPlugin extends AutoPlugin {
...
-            val out   = new File(outR, inR.toURI.relativize(inDir.toURI).getPath) // output dir
-            val cp    = (fullClasspath in Tut).value
-            val opts  = (scalacOptions in Tut).value
-            val pOpts = tutPluginJars.value.map(f => "–Xplugin:" + f.getAbsolutePath)
-            val re    = tutNameFilter.value.pattern.toString
-            tutOne(streams.value, r, in, out, cp, opts, pOpts, re)
-          }
-        }
+      tutOnly := {
+        val in = tutFilesParser(true).parsed
```
Comment (larsrh) [inline plugin/src/main/scala/tut/TutPlugin.scala]: The `true`/`false` parameter looks fishy (could instead be `Space ~> tutFilesParser`). Why does it require the `Space` in the first place?

File: .travis.yml [hunk 3]
```diff
@@ -15,11 +15,11 @@ jdk:
 
 scala:
   - 2.10.6
-  - 2.12.2
+  - 2.12.3
 
 matrix:
   include:
-    - scala: 2.11.8
+    - scala: 2.11.10
```
Comment (stephennancekivell) [inline .travis.yml]: Dont you want 2.11.11 ? I think 2.11.10 had a bug in it.

**Comment by tpolecat** [comment]
Scripted tests will only run for a single Scala version anyway because compiler output changes subtly all the damn time. So really the tests should probably run on sbt 1.0 with Scala 2.12 and we'll call it good. This may be what you have done already but the code is still too blurry to read. I'll check back after more coffee.

File: core/build.sbt [hunk 4]
```diff
@@ -4,9 +4,9 @@ libraryDependencies ++= Seq(
   scalaOrganization.value %  "scala-compiler" % scalaVersion.value
 )
 
-scalaVersion := "2.11.8"
+scalaVersion := "2.12.3"
 
-crossScalaVersions := Seq("2.10.6", scalaVersion.value, "2.12.2")
+crossScalaVersions := Seq("2.10.6", "2.11.10", scalaVersion.value)
```
Comment (SethTisue) [inline core/build.sbt]: 2.11.10 -> 2.11.11

**Comment by larsrh** [comment]
@tpolecat scripted tests are currently running on 2.11 because I couldn't be bothered changing the expected output

**Comment by tpolecat** [comment]
@larsrh 👍 re: scripted

File: .travis.yml (hunk 3 above)
Comment (larsrh) [inline .travis.yml]: Got it, will fix later.

File: core/build.sbt (hunk 4 above)
Comment (larsrh) [inline core/build.sbt]: Got it, will fix later.

File: plugin/src/main/scala/tut/TutPlugin.scala (hunk 1 above)
Comment (dwijnand) [inline plugin/src/main/scala/tut/TutPlugin.scala]: LGTM.

File: plugin/src/main/scala/tut/TutPlugin.scala (hunk 2 above)
Comment (dwijnand) [inline plugin/src/main/scala/tut/TutPlugin.scala]: What is `tutFiles` used for? I can't see anything using it.

**Comment by larsrh** [comment]
@dwijnand Any idea about the failing `update` for 2.10?

File: plugin/src/main/scala/tut/TutPlugin.scala (hunk 2 above)
Comment (larsrh) [inline plugin/src/main/scala/tut/TutPlugin.scala]: If it's unused I can probably axe it.

**Comment by dwijnand** [comment]
I hadn't realised you'd switched the build to 1.0.0-RC3. For context we've been cross-building plugins by bumping the build to sbt 0.13.16 and using `^^`.

I'm not sure what the state of the scripted plugin is, but looks like we stopped releasing 2.10 builds of it for 2.10: http://repo1.maven.org/maven2/org/scala-sbt/scripted-plugin_2.10/

**Comment by larsrh** [comment]
Given that there are source-incompatible changes, I'm not sure how cross-building would work.

**Comment by larsrh** [comment]
Also, why would you need the scripted plugin for 2.10 in the first place? Shouldn't it only require 2.12?

**Comment by tpolecat** [comment]
It's very likely that I set up scripted wrong. You should not presume that anything done here was done for a reason other than "seems to work".

**Comment by tpolecat** [comment]
Due to source incompatibilities I made a new `series/0.6.x` branch that you can target for this PR, if the thingy will let you change it after the fact.

**Comment by larsrh** [comment]
Done.

**Comment by larsrh** [comment]
I've addressed feedback and edited the original PR description to track progress.

File: project/build.properties [hunk 5]
```diff
@@ -1 +1 @@
-sbt.version=0.13.11
+sbt.version=1.0.0-RC3
```
Comment (dwijnand) [inline project/build.properties]: To unblock you I suggest you revert this change and cross-build from 0.13.16.

The scripted-plugin issue you're seeing has been highlighted by @smarter before, but hasn't been addressed yet.

**Comment by dwijnand** [comment]
> cross-build for 0.13.x and 1.0.x (I conjecture this isn't possible)

It is possible from 0.13.16 - it's what we've been doing for all the plugins in http://www.scala-sbt.org/1.x/docs/Community-Plugins.html. Unless you mean something very specific.

File: project/build.properties (hunk 5 above)
Comment (larsrh) [inline project/build.properties]: Ah, I see. I'll update it this evening. Thanks, @dwijnand!

**Comment by larsrh** [comment]
@dwijnand I'm starting to understand now – am I right in assuming that _source compatibility_ between sbt 0.13.x and 1.0.x is not to be expected? Just that the infrastructure for cross building (with sources in different folders) is there?

**Comment by larsrh** [comment]
I'm closing this for now and will continue in separate pull requests. First part: #174.

**Comment by dwijnand** [comment]
@larsrh Correct. But I expect a macro-compat (sbt-compat) to rise. And/or a scalafix-based solution (like catz-cradle was trying to achieve).
